Upgrade to wgpu 0.13

JCapucho opened the pull request:
Upstream `wgpu` introduced some changes to how buffers are mapped, so that now instead of returning a future they expect a callback.

This PR accommodates those changes so that the crate is ready for when 0.13 lands and for easier discoverability for those who want to use wgpu master.

Wumpf commented:
Thank you so much for contributing and even putting up the draft PR before those wgpu changes even landed! 🤩 
I'll have a closer look once those landed and we're ready to publish (:

Imberflur commented:
I noticed wgpu added a new feature that must be enabled for using timestamps inside renderpasses. <https://docs.rs/wgpu/latest/wgpu/struct.Features.html#associatedconstant.WRITE_TIMESTAMP_INSIDE_PASSES>

So this line is no longer completely accurate https://github.com/Wumpf/wgpu-profiler/blob/216ce6ab0881d947448d7705900f2a55fb867e54/src/lib.rs#L42

A few options to address this come to my mind:
* Leave it up to the user to manage where the scopes are placed and what features are required.
* Add the new feature flag to this constant ignoring the case where it may not be available.
* Wire up and expose a toggle for emitting timestamp commands inside of passes similar to the current global toggle. It might be neat to have some way to pass the currently enabled features to `GpuProfiler` and have it automatically configure what is enabled based on that.

cwfitzgerald commented:
I would love the 3rd option - that option won't be available on metal on Apple's M1s, so and this is definitely an important target.

JCapucho commented:
I've updated this PR to point to the crates.io release version, added the missing feature to `REQUIRED_WGPU_FEATURES` and modified the `new` method to automatically configure the profiler behavior as suggested.

Wumpf commented on src/lib.rs:
```diff
@@ -174,7 +176,11 @@ impl GpuProfiler {
+            let mapped_buffers = self.active_frame.mapped_buffers.clone();
+            pool.resolved_buffer_slice().map_async(wgpu::MapMode::Read, move |res| {
+                res.unwrap();
+                mapped_buffers.fetch_add(1, std::sync::atomic::Ordering::SeqCst);
```
SeqCst is a bit too much here. I'm quite sure Ordering::Release is the correct ordering semantic here: We want to ensure that anyone seeing this variable updated (i.e. potentially hitting `frame.query_pools.len()`) will see any other writes (i.e. the buffer being mapped)

pondering actually if relaxed might be enough since pending_frames can't be access in parallel while we're in here, meaning frame.query_pools.len() needs no syncing and the semantics of map should be enough to ensure there is a barrier..... but this gives me a headache so let's not go there ;)

Wumpf commented on src/lib.rs:
```diff
@@ -192,11 +198,7 @@ impl GpuProfiler {
-            .iter_mut()
-            .any(|pool| block_on(poll_once(pool.buffer_mapping.as_mut().unwrap())).is_none())
-        {
+        if frame.mapped_buffers.load(std::sync::atomic::Ordering::SeqCst) != frame.query_pools.len() {
```
should be Ordering::Acquire, rationale see comment above - here we want to acquire the state that got published earlier

Wumpf commented on src/lib.rs:
```diff
@@ -39,12 +36,15 @@ pub struct GpuProfiler {
 #[deny(missing_docs)]
 impl GpuProfiler {
     /// Required wgpu features for timer scopes.
```
we probably shouldn't call it "Required" anymore 🤔

Wumpf replied:
hope @Imberflur isn't too sad if we leave name and doc for it for now :)

Wumpf replied:
ended up "fixing" it in a followup pr

Wumpf reviewed (approved):
Excellent! Agreeing with everyone else here that the "toggles solution" as implemented now is best

Left two comments nerding about atomic ordering, but those don't _really_ matter here. So gonna merge, do that atomics myself (to take the `git blame` on me ;)) and publish a new version
